File: Strivers-A2Z-DSA-Sheet/01.Arrays/3.Hard/python/11.Reverse_pairs.py
"""
QUESTION:
Given an integer array nums, return the number of reverse pairs in the array.
A reverse pair is a pair (i, j) where:
0 <= i < j < nums.length and
nums[i] > 2 * nums[j].

Example:
Input: nums = [1,3,2,3,1]
Output: 2
"""
from typing import List

rev_pair = [0]

def merge(start: int, mid: int, end: int, nums: List[int]) -> None:
    left_size = mid - start + 1
    right_size = end - mid
    left = nums[start:start + left_size]
    right = nums[mid + 1:mid + 1 + right_size]

    # main logic resides here
    m = 0
    for i in range(left_size):
        while m < right_size and left[i] > 2 * right[m]:
            m += 1
        rev_pair[0] += m

    i, j, k = 0, 0, start
    while i < left_size and j < right_size:
        if left[i] > right[j]:
            nums[k] = right[j]
            k += 1
            j += 1
        else:
            nums[k] = left[i]
            k += 1
            i += 1
    while i < left_size:
        nums[k] = left[i]
        k += 1
        i += 1
    while j < right_size:
        nums[k] = right[j]
        k += 1
        j += 1

def mergesort(start: int, end: int, nums: List[int]) -> None:
    if start >= end:
        return
    mid = start + (end - start) // 2
    mergesort(start, mid, nums)
    mergesort(mid + 1, end, nums)
    merge(start, mid, end, nums)

def reversePairs(nums: List[int]) -> int:
    rev_pair[0] = 0
    mergesort(0, len(nums) - 1, nums)
    return rev_pair[0]

"""
TIME COMPLEXITY: O(n log n), where n is the size of the array.
SPACE COMPLEXITY: O(n), where n is the size of the array.
"""

File: Strivers-A2Z-DSA-Sheet/01.Arrays/3.Hard/javascript/11.Reverse_pairs.js
/*
QUESTION:
Given an integer array nums, return the number of reverse pairs in the array.
A reverse pair is a pair (i, j) where:
0 <= i < j < nums.length and
nums[i] > 2 * nums[j].

Example:
Input: nums = [1,3,2,3,1]
Output: 2
*/

let rev_pair = 0;

function merge(start, mid, end, nums) {
    const left_size = mid - start + 1;
    const right_size = end - mid;
    const left = [];
    const right = [];

    for (let i = 0; i < left_size; i++) {
        left[i] = nums[start + i];
    }
    for (let i = 0; i < right_size; i++) {
        right[i] = nums[mid + 1 + i];
    }

    // main logic resides here
    let m = 0;
    for (let i = 0; i < left_size; i++) {
        while (m < right_size && left[i] > 2 * right[m]) {
            m++;
        }
        rev_pair += m;
    }

    let i = 0, j = 0, k = start;
    while (i < left_size && j < right_size) {
        if (left[i] > right[j]) {
            nums[k++] = right[j++];
        }
        else {
            nums[k++] = left[i++];
        }
    }
    while (i < left_size) {
        nums[k++] = left[i++];
    }
    while (j < right_size) {
        nums[k++] = right[j++];
    }
}

function mergesort(start, end, nums) {
    if (start >= end)
        return;
    const mid = start + Math.floor((end - start) / 2);
    mergesort(start, mid, nums);
    mergesort(mid + 1, end, nums);
    merge(start, mid, end, nums);
}

function reversePairs(nums) {
    rev_pair = 0;
    mergesort(0, nums.length - 1, nums);
    return rev_pair;
}

/*
TIME COMPLEXITY: O(n log n), where n is the size of the array.
SPACE COMPLEXITY: O(n), where n is the size of the array.
*/
